src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp

```diff
@@ -182,7 +395,121 @@ namespace
             else value.set(0.0f);
         }
 
+        float compute_pdf(
+            const float             u,
+            const float             v,
+            const float             theta) const
+        {
+            assert(u >= 0.0f && u < 1.0f);
+            assert(v >= 0.0f && v < 1.0f);
+            assert(m_importance_sampler);
+
+            // Compute the probability density of this sample in the importance map.
+            const size_t x = truncate<size_t>(m_importance_map_width * u);
+            const size_t y = truncate<size_t>(m_importance_map_height * v);
+            const float prob_xy = m_importance_sampler->get_pdf(x, y);
+            assert(prob_xy >= 0.0f);
+
+            // Compute the probability density of the emission direction.
+            const float pdf = prob_xy > 0.0f ? prob_xy * m_probability_scale / std::sin(theta) : 0.0f;
+            assert(pdf >= 0.0f);
+
+            return pdf;
+        }
+
+        void build_importance_map(const Project& project, IAbortSwitch* abort_switch)
+        {
+            Stopwatch<DefaultWallclockTimer> stopwatch;
+            stopwatch.start();
+
+            // TODO:
+            // Find importance map width and height from m_inputs (see LatLongMapEnvironmentEDF).
+            m_importance_map_width = 1024;
+            m_importance_map_height = m_importance_map_width / 2;
+
+            m_rcp_importance_map_width = 1.0f / m_importance_map_width;
+            m_rcp_importance_map_height = 1.0f / m_importance_map_height;
+
+            const size_t texel_count = m_importance_map_width * m_importance_map_height;
+            m_probability_scale = texel_count / (2.0f * PiSquare<float>());
+
+            // Create OSL shading system.
+            auto_release_ptr<OIIOTextureSystem> texture_system(
+                OIIOTextureSystemFactory().create());
+
+            RendererServices renderer_services(project, texture_system.ref());
+
+            auto_release_ptr<OSLShadingSystem> osl_shading_system(
+                OSLShadingSystemFactory().create(
+                    &renderer_services,
+                    texture_system.get()));
+
+            // Create shader compiler.
```

Comment:
The shader should have already been compiled and optimized here. You don't need to do it again.

Reply:
OK, I will remove it then. 
For satisfying my curiosity: which kind of optimization is made to shaders?

Reply:
LLVM optimization passes and JIT compiling to x64 assembly.

Reply:
Done, thanks for the explanation.